src/hyp_map.rs

```diff
@@ -2,63 +2,96 @@
 // Licensed under the Apache License, Version 2.0, see LICENSE for details.
 // SPDX-License-Identifier: Apache-2.0
 
+use arrayvec::ArrayVec;
 use page_tracking::{HwMemMap, HwMemRegion, HwMemRegionType, HwReservedMemType, HypPageAlloc};
 use riscv_page_tables::{FirstStagePageTable, PteFieldBits, PteLeafPerms, Sv48};
-use riscv_pages::{InternalClean, Page, PageAddr, PageSize, RawAddr, SupervisorPhys};
+use riscv_pages::{
+    InternalClean, Page, PageAddr, PageSize, RawAddr, SupervisorPhys, SupervisorVirt,
+};
 use riscv_regs::{satp, LocalRegisterCopy, SatpHelpers};
 
-// Returns the base, size, and permission pair for the given region if that region type should be
-// mapped in the hypervisor's virtual address space.
-fn hyp_map_params(r: &HwMemRegion) -> Option<(PageAddr<SupervisorPhys>, u64, PteLeafPerms)> {
-    match r.region_type() {
-        HwMemRegionType::Available => {
-            // map available memory as rwx - unser what it'll be used for.
-            Some((r.base(), r.size(), PteLeafPerms::RWX))
-        }
-        HwMemRegionType::Reserved(HwReservedMemType::FirmwareReserved) => {
-            // No need to map regions reserved for firmware use
+/// Maximum number of regions that will be mapped in the hypervisor.
+const MAX_HYPMAP_REGIONS: usize = 32;
+
+/// Represents a region of the virtual address space of the hypervisor.
+pub struct HypMapRegion {
+    vaddr: PageAddr<SupervisorVirt>,
+    paddr: PageAddr<SupervisorPhys>,
+    page_size: PageSize,
+    page_count: usize,
```

Comment:
So I'm not sure encoding the page size in the region is a good idea -- consider the possible ways to represent a region that is, for example, 5MB and not necessarily 2MB aligned. IMO the easiest way to go about mapping a contiguous region with huge pages would be for `map()` below to pick the appropriate page size based on the alignment and remaining bytes in the region.

Reply:
My expectation was that PageAddr would be aligned to the particular size. So that rather than pushing the decision to use big pages in some inner code would've gone a bit more at higher level.
But can change this.

Reply:
So the issue with that is, each region you add to the list could actually end up being 3 (or more!) separate regions. Since the list is a small-ish fixed size -- and its perfectly possible to do the page-size fitting in the inner loop -- it seems better to avoid dividing regions up by page size in the regions list.

Reply:
Yes I understand. Actually makes the whole thing less architectural so it's probably a good thing.